lib/rucio/core/did.py

```diff
@@ -613,19 +616,34 @@ def delete_dids(dids, account, session=None):
                               'name': did['name']},
                     session=session)
     # Delete rules on did
+    skip_deletion = False  # Skip deletion in case of expiration of a rule
     if rule_id_clause:
         with record_timer_block('undertaker.rules'):
-            for (rule_id, scope, name, rse_expression) in session.query(models.ReplicationRule.id,
-                                                                        models.ReplicationRule.scope,
-                                                                        models.ReplicationRule.name,
-                                                                        models.ReplicationRule.rse_expression).filter(or_(*rule_id_clause)):
+            for (rule_id, scope, name, rse_expression, locks_ok_cnt, locks_replicating_cnt, locks_stuck_cnt) in session.query(models.ReplicationRule.id,
+                                                                                                                              models.ReplicationRule.scope,
+                                                                                                                              models.ReplicationRule.name,
+                                                                                                                              models.ReplicationRule.rse_expression,
+                                                                                                                              models.ReplicationRule.locks_ok_cnt,
+                                                                                                                              models.ReplicationRule.locks_replicating_cnt,
+                                                                                                                              models.ReplicationRule.locks_stuck_cnt).filter(or_(*rule_id_clause)):
                 logging.debug('Removing rule %s for did %s:%s on RSE-Expression %s' % (str(rule_id), scope, name, rse_expression))
+
                 # Propagate purge_replicas from did to rules
                 if (scope, name) in not_purge_replicas:
                     purge_replicas = False
                 else:
                     purge_replicas = True
-                rucio.core.rule.delete_rule(rule_id=rule_id, purge_replicas=purge_replicas, delete_parent=True, nowait=True, session=session)
+                if expire_rules and locks_ok_cnt + locks_replicating_cnt + locks_stuck_cnt > 10000:
```

Comment:
Make it configurable in a section undertaker in the config table ?

Reply:
I am merging this now, as it should be in todays release. The configuration will come with #1932